File: matlab/orth/orth_dcgs2_arnoldi_cleanup.m
%
   function [ q, h ] = orth_dcgs2_arnoldi_cleanup( Q, H )
%  
      m = size(Q,1);
      k = size(Q,2); 
%
      work = zeros(k,1);
      h = zeros(k,1);
      q = zeros(m,1);
%
      work(1:k-1,1) = Q(1:m,1:k-1)' * Q(1:m,k);
      work(k,1)  = Q(1:m,k)' * Q(1:m,k);         
      h(k,1) = sqrt( work(k,1) - work(1:k-1,1)' * work(1:k-1,1) ); 
      q(1:m,1) = Q(1:m,k) - Q(1:m,1:k-1) * work(1:k-1,1);
      q(1:m,1) = q(1:m,1) / h(k,1);
      h(1:k-1,1) = H(1:k-1,k-1) + work(1:k-1,1);
%
   end 

File: julia/orth/orth_dcgs2_arnoldi_cleanup.jl
using LinearAlgebra

function orth_dcgs2_arnoldi_cleanup(Q::AbstractMatrix, H::AbstractMatrix)
    m = size(Q, 1)
    k = size(Q, 2)
    work = zeros(k)
    h = zeros(k)

    work[1:k-1] = Q[:, 1:k-1]' * Q[:, k]
    work[k]     = Q[:, k]' * Q[:, k]
    h[k] = sqrt(work[k] - work[1:k-1]' * work[1:k-1])
    q = (Q[:, k] .- Q[:, 1:k-1] * work[1:k-1]) ./ h[k]
    h[1:k-1] = H[1:k-1, k-1] .+ work[1:k-1]

    return q, h
end
